### ISSUE TITLE
Unset `google.protobuf.StringValue` map values are decoded as empty strings instead of empty values

### ISSUE THREAD

**Issue by ergl** [tomas-abrahamsson/gpb#216]
I'm having trouble making `google.protobuf.StringValue` work between two services I have, one in Erlang and another in Go.

Here's the minimal proto definition:

```protobuf
syntax = "proto3";
package client;

import "google/protobuf/wrappers.proto";

message Foo {
    map<uint32, google.protobuf.StringValue> fields = 1;
}
```

The idea here is that I want to set some `uint32` keys to a string, while having the ability of setting others to an empty value. On the Go side, I put the string "value0" for key `0`, and `nil` for key `1`.

When decoding this message on the Erlang side, I get the following:

```erlang
#{fields =>
      #{0 => #{value => <<"value0">>},1 => #{value => <<>>}}
```

Is there a way to get some marker value for the key `1` so that I know that the sender decided to leave the key empty?

Conversely, I can't find a way to send from Erlang to Go something like `#{0 => #{value => <<"value0">>}, 1 => #{}}`, and get it decoded as `0 -> "value0"; 1 -> nil` on the Go side (it currently decodes as `0 -> "value0"; 1 -> ""`)

**Comment by ergl** [comment]
Forgot to say. On the Erlang side, I would expect my first example to decode as follows:

```erlang
#{fields =>
      #{0 => #{value => <<"value0">>},1 => #{}} % Empty map for key 1
```

**Comment by ergl** [comment]
Actually, it seems that the Go version of protocol buffers behaves in the same way. So it seems like this is the expected behavior. I find it strange and unintuitive, though.

**Comment by tomas-abrahamsson** [comment]
Hi, yes, with maps, it works like that. I checked this earlier: Your `map<uint32, google.protobuf.StringValue>` is [said](https://developers.google.com/protocol-buffers/docs/proto3#backwards_compatibility) to be equivalent _on the wire_ to a repeated field MapElement defined like this: `message MapElement { uint32 key = 1; google.protobuf.StringValue value = 2; }` And then your question is akin to whether the `value` field may be unset to indicate `nil`. I tried this out a while back with C++ [in issue 111](https://github.com/tomas-abrahamsson/gpb/issues/111#issuecomment-315284566), but it appears not to work that way, unfortunately. And I suppose the Go implementation works the same as the one in C++. (Even though that comment was in the context of `proto2`, it seems to work as if it had been `proto3` for map type fields, so there is no difference in this respect.)

(Had it worked that way you first thought, then one could have implemented a set type as a map with set elements as keys and omitted values. However, on the other hand, a map with omitted keys and only values would have been kind of odd.)
